tests/shouldwork/Cores/Xilinx/XpmCdcHandshakeTypes.hs

```diff
@@ -0,0 +1,128 @@
+
+import Clash.Cores.Xilinx.Xpm.Cdc.Handshake
+import Clash.Explicit.Prelude
+import Clash.Explicit.Testbench
+import Data.Proxy
+import Language.Haskell.TH.Lib
+
+createDomain vXilinxSystem{vName="D3",  vPeriod=hzToPeriod 30e6}
+createDomain vXilinxSystem{vName="D5",  vPeriod=hzToPeriod 50e6}
+createDomain vXilinxSystem{vName="D10", vPeriod=hzToPeriod 100e6}
+createDomain vXilinxSystem{vName="D11", vPeriod=hzToPeriod 110e6}
+
+data State = WaitForDeassert | WaitForAssert deriving (Generic, NFDataX)
+
+noRst :: KnownDomain dom => Reset dom
+noRst = unsafeFromHighPolarity (pure False)
+
+-- | Transfer 1, 2, 3, ... to destination domain
+srcFsm ::
+  forall a src .
+  ( KnownDomain src
+  , Num a
+  , NFDataX a
+  ) =>
+  Clock src ->
+  Signal src Bool ->
+  Signal src (a, Bool)
+srcFsm clk = mealy clk noRst enableGen go (0, WaitForDeassert)
+ where
+  go (n, WaitForDeassert) True  = ((n,     WaitForDeassert), (n,     False))
+  go (n, WaitForDeassert) False = ((n + 1, WaitForAssert),   (n + 1, True))
+  go (n, WaitForAssert)   False = ((n,     WaitForAssert),   (n,     True))
+  go (n, WaitForAssert)   True  = ((n,     WaitForDeassert), (n,     False))
+{-# NOINLINE srcFsm #-}
+
+-- | Receives data from source domain
+dstFsm ::
+  forall a dst .
+  KnownDomain dst =>
+  Clock dst ->
+  Signal dst (Bool, a) ->
+  Signal dst (Bool, Maybe a)
+dstFsm clk = mealy clk noRst enableGen go WaitForAssert
+ where
+  go WaitForAssert   (False, _) = (WaitForAssert,   (False, Nothing))
+  go WaitForAssert   (True,  n) = (WaitForDeassert, (True,  Just n))
+  go WaitForDeassert (True,  _) = (WaitForDeassert, (True,  Nothing))
+  go WaitForDeassert (False, _) = (WaitForAssert,   (False, Nothing))
+{-# NOINLINE dstFsm #-}
+
+-- | Composition of 'srcFsm' and 'dstFsm'
+top ::
+  forall a srcStages dstStages src dst .
+  ( KnownDomain src
+  , KnownDomain dst
+  , Num a
+  , NFDataX a
+  , BitPack a
+  , 1 <= BitSize a, BitSize a <= 1024
+  , 2 <= srcStages, srcStages <= 10
+  , 2 <= dstStages, dstStages <= 10
+  ) =>
+  XpmCdcHandshakeConfig srcStages dstStages ->
+  Clock src ->
+  Clock dst ->
+  Signal dst (Maybe a)
+top opts clkSrc clkDst = maybeDat
+ where
+  (srcIn, srcSend) = unbundle $ srcFsm @a clkSrc srcRcv
+
+  (destOut, destReq, srcRcv) =
+    xpmCdcHandshakeWith opts clkSrc clkDst srcIn srcSend destAck
+
+  (destAck, maybeDat) =
+    unbundle $ dstFsm @a clkDst $ bundle (destReq, destOut)
+{-# NOINLINE top #-}
+
+tb ::
+  forall a b srcStages dstStages n .
+  ( KnownNat n, 1 <= n
+  , 2 <= srcStages, srcStages <= 10
+  , 2 <= dstStages, dstStages <= 10
+  , KnownDomain a
+  , KnownDomain b
+  ) =>
+  Proxy a -> Proxy b ->
+  XpmCdcHandshakeConfig srcStages dstStages ->
+  -- | Expected data (Maybe + data)
+  Vec n (BitVector 9) ->
+  Signal b Bool
+tb Proxy Proxy opts expectedDat = done
+ where
+  actual = top @(Unsigned 8) opts clkA clkB
+
+  done =
+    outputVerifierWith
+      (\clk rst -> assertBitVector clk rst "outputVerifier Port A")
```

Comment:
Since there are no undefined samples, how about using regular `assert`? Actually, regular `outputVerifer` would do as there is only a single `assert` in the test bench, so it's not important what its label is. I kinda forgot about that factoid in previous reviews of these XPM's, it seems rather laborious to use `outputVerifierWith` if you don't have multiple `assert`s anyway.

Reply:
The tangential advantage of `assert` over `assertBitVector` is its readability.

Reply:
Ah. But this only fails in HDL simulation, by definition, and there the readability of a `Maybe (Unsigned 8)` is not going to be great no matter which `assert`.

Reply:
As discussed, this now uses regular ol' `outputVerifier'`.